### PR TITLE
Proposing a new RecipeReference model, to gradually migrate

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Omit
Docs: Omit

Some design ideas:

- Make it mutable, so it can be centralized and shared, instead of passed by value and copied
- Validation should probably be orthogonal, and applied only at certain critical points
- New model should provide an incremental migration, with conversion to/from the legacy model
- The new model should enable the new use syntax, like using ``pkg/version`` instead of ``pkg/version@``.
- New model should be clearly differentiating "requirements", like ``pkg/[>1.0]`` including version ranges, from actual references, which use fixed, defined versions.
- Integrate timestamps in the model
- Allow ``pkg/version@user`` definitions, without channel
- Try to define an integrated string format for timestamps ``pkg/version@user/channel#rev%timestamp``. Being timestamp an integer. Or should we use human readable dates for text serialization always?

To be processed after @lasote Python API proposals, this could be an object that maybe could be in the API interface.

### CODE REVIEW

File: conans/model/ref.py [hunk 1]
```diff
@@ -146,6 +146,91 @@ def validate_revision(revision):
...
+        return result
+
+    def __lt__(self, ref):
+        # The timestamp goes before the revision for ordering revisions chronologically
+        return (self.name, self.version, self.user, self.channel, self.timestamp, self.revision) \
+               < (ref.name, ref.version, ref.user, ref.channel, ref.timestamp, ref.revision)
+
+    def __eq__(self, other):
+        # TODO: In case of equality, should it use the revision and timestamp?
+        raise Exception("WHO IS COMPARING RECIPE REFERENCES?")
```
Comment (lasote) [inline conans/model/ref.py]: LOL

File: conans/model/ref.py [hunk 2]
```diff
@@ -146,6 +146,91 @@ def validate_revision(revision):
...
+        # return not self.__eq__(other)
+
+    def __hash__(self):
+        raise Exception("WHO IS COMPARING RECIPE REFERENCES?")
+        # return hash((self.name, self.version, self.user, self.channel, self.revision))
+
+    @staticmethod
+    def loads(text):  # TODO: change this default to validate only on end points
+        try:
+            # revision
```
Comment (lasote) [inline conans/model/ref.py]:
```suggestion
            # timestamp
```

**Review by lasote — COMMENTED** [review]
We could start introducing it, I don't see a problem.

File: conans/model/ref.py (hunk 1 above)
Comment (memsharded) [reply conans/model/ref.py]: Comparing by equality default comparison has bit me more times that I would like 😂

**Comment by czoido** [comment]
Looks good

**Review by czoido — APPROVED** [review]

**Pull request merged**